fabric-item-api-v1/src/main/java/net/fabricmc/fabric/api/client/item/v1/ArmorTextureRegistry.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.client.item.v1;
+
+import java.util.Arrays;
+import java.util.Objects;
+
+import net.minecraft.entity.EquipmentSlot;
+import net.minecraft.entity.LivingEntity;
+import net.minecraft.item.ItemConvertible;
+import net.minecraft.item.ItemStack;
+
+import net.fabricmc.api.EnvType;
+import net.fabricmc.api.Environment;
+import net.fabricmc.fabric.impl.client.item.ArmorTextureRegistryImpl;
+
+/**
+ * A class for registering custom armor textures for {@link ItemConvertible}, to be provided by a {@link Provider}.
+ *
+ * <p>
+ * This can be used to replace existing vanilla armor textures conditionally, however each {@link net.minecraft.item.Item}
+ * instance can only allow one {@link Provider}.
+ * </p>
+ *
+ * <p>
+ * Since armor textures identifier in vanilla is hardcoded to be in the {@code minecraft} namespace, this registry can also be
+ * used to use a custom namespace if desired.
+ * </p>
```

Comment:
This needs more documentation on how models and textures relate to each other.

The suggestion around replacing vanilla armor should be more like a warning along the lines of "While this can be used to replace vanilla armor, there can be only one replacement at a time, causing potential conflicts with other mods. Consider using a separate item instead." 

Armor models and texture are so tightly connected that a single ArmorRenderingRegistry class may be more appropriate.

IMO we should also offer a wider hook that effectively replaces all of `renderArmor()` with a handler for more invasive customization.

Reply:
> IMO we should also offer a wider hook that effectively replaces all of renderArmor() with a handler for more invasive customization.
The BipedEntityModel _should be_ possible to be entirely customized, you can completely rip out BipedEntityModel#render.